### ISSUE TITLE
dotnet Core for Raspberry PI (Buster)

### ISSUE THREAD

**Issue by w4mmp** [dotnet/core#4520]
No .Net Framework for in most current release of .Net Core for the ARM platform (Raspberry PI).

Hello,
I am new to the .Net Core environment for the ARM platform.  I am not new to .Net Framework.  I have a .Net Framework application that has been developed on the Windows Platform.  I am attempting to run it on a Raspberry PI 4.   The most current .Net Core runtime and SDK have been installed.   
ron@multus-sdr:~/mscc/gui/dotnet $ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.102
 Commit:    573d158fea

Runtime Environment:
 OS Name:     raspbian
 OS Version:  10
 OS Platform: Linux
 RID:         linux-arm
 Base Path:   /home/ron/src/dotnet-arm32/sdk/3.1.102/

Host (useful for support):
  Version: 3.1.2
  Commit:  916b5cba26

.NET Core SDKs installed:
  3.1.102 [/home/ron/src/dotnet-arm32/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.2 [/home/ron/src/dotnet-arm32/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.2 [/home/ron/src/dotnet-arm32/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

When attempting to run the .Net Framework application on the RPi, I receive this:

ron@multus-sdr:~/mscc/gui/dotnet $ dotnet mscc.dll
It was not possible to find any compatible framework version
The framework 'Microsoft.WindowsDesktop.App', version '3.1.0' was not found.
  - No frameworks were found.

You can resolve the problem by installing the specified framework and/or SDK.

The specified framework can be found at:
  - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.WindowsDesktop.App&framework_version=3.1.0&arch=arm&rid=raspbian.10-arm
ron@multus-sdr:~/mscc/gui/dotnet $

So the question is:   Will .Net Framework be made available?

Regards,
Ron

**Comment by scalablecory** [comment]
Is the application built for .NET Framework, or .NET Core? Apps built for one will not run on the other.

**Comment by scalablecory** [comment]
Also WPF and System.Windows.Forms are Windows-only. If your app was built against one of those, it will not work in Linux.

**Comment by leecow** [comment]
Only .NET Core is supported on ARM devices. Downloads for the SDK or Runtime binary archives (tar.gz) can be found at https://dotnet.microsoft.com/download/dotnet-core/3.1 (ARM32 for Pi 3 and ARM64 for Pi4). Instructions for installing using the tar.gz file are in the latest [Install Instructions](https://github.com/dotnet/core/blob/master/release-notes/3.1/3.1.3/3.1.3-install-instructions.md#installation-from-a-binary-archive) document.

**Comment by w4mmp** [comment]
Hello,

To address this and other responses.   The application is a .Net Core application with Microsoft.Windows.Compatibility installed so that I have access to .Net Framework API.   This works on Windows but does not on Linux ARM.

The question is:  Will  .Net Core support .Net Framework on Linux in the future?

Regards,

Ron Patton

**Comment by RobbieZA** [comment]
I have the same issue.
I have a Raspberry Pi 4b running Ubuntu 19.10 ARM64 vesion:
  No LSB modules are available.
  Distributor ID:	Ubuntu
  Description:	Ubuntu 19.10
  Release:	19.10
  Codename:	eoan

I am wanting to host the sample Blazor site and have followed all the instructions (https://docs.microsoft.com/en-us/dotnet/core/install/linux-package-manager-ubuntu-1910#troubleshoot-the-package-manager) to install ASP.NET Core Runtime 3.1.3 and .NET Core Runtime 3.1.3 (both ARM64 versions) manually. (wget  and unzipped with tar zxf).
When I try: dotnet --list-sdks in the specific folder I get command not found.

Any help will be very much appreciated.

**Comment by mairaw** [comment]
@joperezr can you help with questions re. setup on Raspberry PI?

@w4mmp the Microsoft.Windows.Compatibility package contains a lot of APIs that are Windows only. It's meant to make it easier to port your .NET Framework into .NET Core. More info can be found at https://docs.microsoft.com/en-us/dotnet/core/porting/windows-compat-pack.

But to answer your question, .NET Core won't support .NET Framework on Linux.

**Comment by joperezr** [comment]
Hello @w4mmp as @mairaw points out .NET Framework is not supported on Linux, although many .NET Framework applications are able to run in Linux using Mono. That said, what you really want is to move to .Net Core, since it does support Linux, does support arm architecture (inclidding Raspberry Pi devices) and supports most of the API that .Net Framework has plus many more. All that said, as @scalablecory suggests, WPF and winforms don’t really work in Linux so if your app depends on that, then it won’t really work.

My suggestion for you since you are already a .NET Framework developer, would be to download the .Net Core sdk on your windows machine (since you are already familiar with developing in Windows), create a new console app ( you can do this by simply running `dotnet new console`) and open the project that gets created with Visual Studio (or the IDE of your preference). After that you can either build just a hello world app, or copy the code over from your .net framework app, and once you are done, you should run the following command from the command prompt `dotnet publish -r linux-arm`. This will create a publish folder that you can copy directly to your raspberry pi Which should just be able to Run it by invoking the executable that got placed (which will have the name of your project) on that Publish folder. You won’t need to install .Net Core on the Pi if you don’t want to, as publishing the way I suggested will create a self contained application which will have all dependencies deployed with your app.

Let us know if you hit any issues whole you start playing with that 😄

**Comment by joperezr** [comment]
@RobbieZA not sure how you downloaded dotnet on your Pi but it sounds like you probably didn’t add the path where you installed it to the PATH.

Can you try these instructions instead?
```
wget https://download.visualstudio.microsoft.com/download/pr/15434752-2803-4722-b406-9a62d3e089d5/159bd7713f8a7ff82d7b2f756ae87f4f/dotnet-sdk-3.1.202-linux-arm64.tar.gz
mkdir -p $HOME/dotnet && tar zxf dotnet-sdk-3.1.202-linux-arm64.tar.gz -C $HOME/dotnet
export DOTNET_ROOT=$HOME/dotnet
export PATH=$PATH:$HOME/dotnet
```

**Comment by joperezr** [comment]
@RobbieZA also do note that as I suggested to @w4mmp you can do the development in Windows if you prefer and then just deploy into the raspberry pi once your app is ready by deploying a self-contained app. This way you won’t need to install the sdk on the raspberry pi itself. I usually recommend this as the processors on most PCBs are not very powerful so usually developing on them is slower.

**Comment by w4mmp** [comment]
Hello,
Thanks for the info but I need to address one aspect of this.   I have a heavy investment in my .net framework application.  Which is why I approached the subject of .net core and framework running on the Raspberry Pi.  The application consists of many thousand lines of C#,  framework code.   I have ported the application to .net core (under Windows) with the framework extensions as an exercise to determine how difficult it would be.  That worked.   But the point of the port was with eventual target of the Raspberry Pi in mind.   This as I now know, is not possible.   I have tried running the application under Mono and found Mono to be severely lacking in a number aspects.  Also, support is almost nonexistent for Mono.    In retrospect I should have written the application with the services of Qt which is truly platform independent.   I find it a shame there is no effort or direction to support framework on an extremely popular platform such as the Raspberry Pi.  Particularly since the 4B version is available that has plenty of horse power to run GUI applications.

**Comment by joperezr** [comment]
I might be misunderstanding here, but I believe your feedback is very specific towards support for wpf and winforms to run in Linux, not really .Net Framework, as .Net Core already supports most of the Apis in .Net Framework (plus much more) and it also comes with special shims that will allow you to load .NET Framework assemblies into a .Net Core application. For wpf and winforms case, unfortunately the implementation is very dependent on windows OS, which will make it impossible to get it to work on other platforms. That said, what I’ve seen a lot of people do in your situation is to try to decouple the logic of your app from the GUI as much as possible and then implement the GUI layer using winforms/wpf when running on Windows and using other alternatives like Web or Xamarin when running on other platforms with both implementations using the same shared main logic layer underneath.

**Comment by mairaw** [comment]
Also, not sure if you've tried running the .NET Portability Analyzer to see how much your current project is compatible with .NET Core and what are the problematic APIs:
https://docs.microsoft.com/en-us/dotnet/standard/analyzers/portability-analyzer

This would give you an idea of the effort to port.

**Comment by w4mmp** [comment]
Hi joperezr,
Decouple the GUI?  It is already decoupled.   The overall architecture of the application is a true client/sever.  All processing occurs in back end servers.  The GUI is there only to handle the user interface.  Communications between the GUI and the servers is via UDP.  The back end servers have already been ported to the Raspberry Pi.  So at the moment,  the servers are running on the RPi and the GUI is running on Windows.  However the GUI is quite involved and as I have mentioned there are many thousands of line of code supporting the GUI.   

Hi mairaw,
Yes I have been down the path of the .Net Portability Analyzer.  The results were that the GUI is compatible with .Net Core.

The GUI should run on Linux IF the hardware platform is x86.   But that is not the target.  The target is Linux on the Raspberry Pi.  That is where things fall apart.   While .Net Core is supported on the RPi,  .Net Core with the .Net Framework extension is not.   The bottom line is the .Net Framework is not supported on the RPi and it appears there is no plan in place for that to happen.   My only option is to try and get the GUI to run under Mono (which is not promising) or do a complete re-write of the GUI utilizing a development platform that is friendly to the RPi.  

Regards,
Ron

**Comment by RobbieZA** [comment]
> @RobbieZA not sure how you downloaded dotnet on your Pi but it sounds like you probably didn’t add the path where you installed it to the PATH.
> 
> Can you try these instructions instead?
> ```
> wget https://download.visualstudio.microsoft.com/download/pr/15434752-2803-4722-b406-9a62d3e089d5/159bd7713f8a7ff82d7b2f756ae87f4f/dotnet-sdk-3.1.202-linux-arm64.tar.gz
> mkdir -p $HOME/dotnet && tar zxf dotnet-sdk-3.1.202-linux-arm64.tar.gz -C $HOME/dotnet
> export DOTNET_ROOT=$HOME/dotnet
> export PATH=$PATH:$HOME/dotnet
> ```

Thank you for the info, I have done a clean install of Ubuntu 20.04 64-bit. I will run as stand alone, do you recommend running on LAMP or MGINX. The plan is to test having many API calls from multiple users (around 100) and SQLlite?

**Comment by w4mmp** [comment]
73,
Ron / W4MMP

**Comment by Thraka** [comment]
No one has asked you specifically, what GUI are you using? We're assuming it's WinForms or WPF. Either way, those aren't supported on any operating system other than Windows as they have deep dependencies within Windows itself.

As .NET Core is cross-platform and .NET Framework is Windows-only. The Windows Compatibility library exposes a lot of Windows-only APIs that .NET Core doesn't have by default. This makes it easy to port your Windows-only application from .NET Framework to .NET Core, but requires you to remain Windows-only as those APIs are still Windows-only.

While WinForms/WPF support for .NET Core is available, they are still GUI frameworks that lean on Windows and therefore can only run on Windows. It's just having a C++ app that can compile and run on multiple platforms, but if you have a dependency on a library only available on Linux, that app simply cannot run correctly on Windows.

.NET 6 is introducing a cross-platform GUI technology called .NET MAUI. You can read about it here: https://devblogs.microsoft.com/dotnet/introducing-net-multi-platform-app-ui/ In brief though, this is to be said about .NET MAUI:

>.NET MAUI is an evolution of the increasingly popular Xamarin.Forms toolkit ... For years companies such as UPS, Ernst & Young, and Delta have been leveraging the mobile expertise of Xamarin atop .NET to power their businesses ... .NET MAUI extends this success on mobile to embrace the desktop making it the best way to build multi-platform applications

Otherwise, there are some other technologies you can look at while making your app GUI cross-platform, such as [Eto](https://github.com/picoe/Eto) and [Avalonia](https://github.com/AvaloniaUI/Avalonia).

Obviously porting an app GUI to a whole other framework is a lot of work. I think with this information what you're really asking for is for the GUI framework you're using, which is Windows-only, to be ported to other operating systems, yes? If so, you should formalize your request with specifying those specific details in your original post here. However, I think there are requests already open related to [wpf](https://github.com/dotnet/wpf/issues/48) and [winforms](https://github.com/dotnet/winforms/issues/3383), and I think generally the feedback is the same, it's not a goal. With .NET MAUI coming, which uses XAML technology, I think that is something to watch for.

**Comment by RobbieZA** [comment]
> @RobbieZA not sure how you downloaded dotnet on your Pi but it sounds like you probably didn’t add the path where you installed it to the PATH.
> 
> Can you try these instructions instead?
> 
> ```
> wget https://download.visualstudio.microsoft.com/download/pr/15434752-2803-4722-b406-9a62d3e089d5/159bd7713f8a7ff82d7b2f756ae87f4f/dotnet-sdk-3.1.202-linux-arm64.tar.gz
> mkdir -p $HOME/dotnet && tar zxf dotnet-sdk-3.1.202-linux-arm64.tar.gz -C $HOME/dotnet
> export DOTNET_ROOT=$HOME/dotnet
> export PATH=$PATH:$HOME/dotnet
> ```

I loaded the DotNet Core Runtime on my Raspberry Pi 4 4Gb Running Ubuntu 20.04 and am running a Web API to service an App written in Xamarin. Works really well. Now to stress test and see what the combination is capable of.

**Comment by JakeSays** [comment]
@RobbieZA what UI framework is xamarin using on the PI?

**Comment by JakeSays** [comment]
@w4mmp You're trying to port a Windows .NET GUI application to linux?

**Comment by RobbieZA** [comment]
No the Xamarin app (forms - Android) is making HTTP requests to the webapi running on the RPi4.
